nornir/core/inventory.py

```diff
@@ -178,91 +212,54 @@ def get(self, item, default=None):
         except KeyError:
             return default
 
-    @property
-    def nornir(self):
-        """Reference to the parent :obj:`nornir.core.Nornir` object"""
-        return self._nornir
-
-    @nornir.setter
-    def nornir(self, value):
-        # If it's already set we don't want to set it again
-        # because we may lose valuable information
-        if not getattr(self, "_nornir", None):
-            self._nornir = value
-
-    @property
-    def hostname(self):
-        """String used to connect to the device. Either ``hostname`` or ``self.name``"""
-        return self.get("hostname", self.name)
-
-    @hostname.setter
-    def hostname(self, value):
-        self.data["hostname"] = value
-
-    @property
-    def port(self):
-        """Either ``port`` or ``None``."""
-        return self.get("port")
-
-    @port.setter
-    def port(self, value):
-        self.data["port"] = value
-
-    @property
-    def username(self):
-        """Either ``username`` or user running the script."""
-        return self.get("username", getpass.getuser())
-
-    @username.setter
-    def username(self, value):
-        self.data["username"] = value
-
-    @property
-    def password(self):
-        """Either ``password`` or empty string."""
-        return self.get("password", "")
-
-    @password.setter
-    def password(self, value):
-        self.data["password"] = value
-
-    @property
-    def platform(self):
-        """OS the device is running. Defaults to ``platform``."""
-        return self.get("platform")
-
-    @platform.setter
-    def platform(self, value):
-        self.data["platform"] = value
-
     def get_connection_parameters(
         self, connection: Optional[str] = None
-    ) -> Dict[str, Any]:
+    ) -> ConnectionOptions:
         if not connection:
-            return {
-                "hostname": self.hostname,
-                "port": self.port,
-                "username": self.username,
-                "password": self.password,
-                "platform": self.platform,
-                "connection_options": {},
-            }
+            d = ConnectionOptions(
+                hostname=self.hostname,
+                port=self.port,
+                username=self.username,
+                password=self.password,
+                platform=self.platform,
+                extras={},
+            )
         else:
-            conn_params = self.get(f"{connection}_options", {})
-            return {
-                "hostname": conn_params.get("hostname", self.hostname),
-                "port": conn_params.get("port", self.port),
-                "username": conn_params.get("username", self.username),
-                "password": conn_params.get("password", self.password),
-                "platform": conn_params.get("platform", self.platform),
-                "connection_options": conn_params.get("connection_options", {}),
-            }
+            r = self._get_connection_options_recursively(connection)
```

Comment:
This get connection_parameters doesn't look to be working properly. For example, with this hosts.yaml and groups.yaml:

```
# hosts.yaml
---
nxos1:
    hostname: nxos1.twb-tech.com
    groups: 
      - nxos
```

```
# groups.yaml
---
defaults:
  username: admin
  password: foo
  port: 22

nxos:
  port: 8443
  connection_options:
    napalm:
      platform: nxos
      extras:
        timeout: 300
        optional_args:
          port: 8443
  data:
    foo: whatever
```

Here is what the parameters look like in the napalm connection plugin right after the `parameters.update(extras)`

```python
(Pdb) p parameters
{'hostname': None, 'username': None, 'password': None, 'optional_args': {'port': 8443}, 'timeout': 300}

(Pdb) p platform
'nxos'
```

I stepped through this code in Pdb and it it looked like it largely grabbed the attributes from the first group and just returned that.

Note, my tests have the following change applied from our discussion yesterday:

```python
diff --git a/nornir/core/inventory.py b/nornir/core/inventory.py
index b3a4709..4ea2e28 100644
--- a/nornir/core/inventory.py
+++ b/nornir/core/inventory.py
@@ -178,7 +178,7 @@ class Host(InventoryElement):
         v = object.__getattribute__(self, name)
         if v is None:
             for g in self.groups.refs:
-                r = object.__getattribute__(self, name)
+                r = object.__getattribute__(g, name)
                 if r is not None:
                     return r
```

Reply:
Mmm, that's odd. Let me go through this tomorrow. Thanks for the thorough review.

Reply:
I tried it both with and without the one above diff and saw the same thing (just to verify my code change didn't break anything).

Reply:
Ok, I **think** I solved your problem but I am not sure as I was unsure how to replicate it. You clearly pointed me to a bug (which was slightly more involved that we initially thought). Also, note that defaults, are now under `defaults.yaml`:

https://github.com/nornir-automation/nornir/blob/inventory_rework/tests/conftest.py#L56